internal/transport/http2_client.go

```diff
@@ -1254,14 +1256,122 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
-	if h2code, err := state.decodeHeader(frame); err != nil {
-		t.closeStream(s, err, true, h2code, status.Convert(err), nil, endStream)
+	// frame.Truncated is set to true when framer detects that the current header
+	// list size hits MaxHeaderListSize limit.
+	if frame.Truncated {
+		se := status.New(codes.Internal, "peer header list size exceeded limit")
+		t.closeStream(s, se.Err(), true, http2.ErrCodeFrameSize, se, nil, endStream)
 		return
 	}
 
+	var (
+		// If a gRPC Response-Headers has already been received, then it means that the peer is speaking gRPC and we are in gRPC mode.
+		isGRPC         = !initialHeader
+		mdata          = make(map[string][]string)
+		contentTypeErr string
+		grpcMessage    string
+		grpcStatus     *status.Status
+		statusGen      *status.Status
+		rawStatusCode  = codes.Unknown
+	)
+
+	for _, hf := range frame.Fields {
+		switch hf.Name {
+		case "content-type":
+			if _, validContentType := grpcutil.ContentSubtype(hf.Value); !validContentType {
+				contentTypeErr = fmt.Sprintf("transport: received the unexpected content-type %q", hf.Value)
+				break
+			}
+			// TODO: do we want to propagate the whole content-type in the metadata,
+			// or come up with a way to just propagate the content-subtype if it was set?
+			// ie {"content-type": "application/grpc+proto"} or {"content-subtype": "proto"}
+			// in the metadata?
+			mdata[hf.Name] = append(mdata[hf.Name], hf.Value)
+			isGRPC = true
+		case "grpc-encoding":
+			s.recvCompress = hf.Value
+		case "grpc-status":
+			code, err := strconv.Atoi(hf.Value)
+			if err != nil {
+				grpcStatus = status.New(codes.Internal, fmt.Sprintf("transport: malformed grpc-status: %v", err))
+				break
+			}
+			rawStatusCode = codes.Code(uint32(code))
+		case "grpc-message":
+			grpcMessage = decodeGrpcMessage(hf.Value)
+		case "grpc-status-details-bin":
+			v, err := decodeBinHeader(hf.Value)
+			if err != nil {
+				grpcStatus = status.New(codes.Internal, fmt.Sprintf("transport: malformed grpc-status-details-bin: %v", err))
+				break
+			}
+			st := &spb.Status{}
+			if err := proto.Unmarshal(v, st); err != nil {
+				grpcStatus = status.New(codes.Internal, fmt.Sprintf("transport: malformed grpc-status-details-bin: %v", err))
+				break
+			}
+			statusGen = status.FromProto(st)
+		case ":status":
+			// we only process :status if this is not grpc so no need to ALWAYS convert the value
+			mdata[hf.Name] = append(mdata[hf.Name], hf.Value)
+		case "grpc-tags-bin":
+			v, err := decodeBinHeader(hf.Value)
+			if err != nil {
+				grpcStatus = status.New(codes.Internal, fmt.Sprintf("transport: malformed grpc-tags-bin: %v", err))
+				break
+			}
+			mdata[hf.Name] = append(mdata[hf.Name], string(v))
+		case "grpc-trace-bin":
+			v, err := decodeBinHeader(hf.Value)
+			if err != nil {
+				grpcStatus = status.New(codes.Internal, fmt.Sprintf("transport: malformed grpc-trace-bin: %v", err))
+				break
+			}
+			mdata[hf.Name] = append(mdata[hf.Name], string(v))
```

Comment:
These two are essentially the same.  Please combine them (`case "grpc-tags-bin", "grpc-trace-bin":`)

Reply:
Actually, I _think_ these can be handled by the `default` case, too, since `decodeMetadataHeader` handles `-bin` suffixes.  Let's do that then if it works.

Reply:
These can be handled by the default case, and it definitely cleans up the code a bit!